src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Diagnostics/RazorTranslateDiagnosticsService.cs

```diff
@@ -162,23 +163,34 @@ private Diagnostic[] FilterCSharpDiagnostics(Diagnostic[] unmappedDiagnostics, R
     private static bool InCSharpLiteral(
         Diagnostic d,
         SourceText sourceText,
-        RazorSyntaxTree syntaxTree,
-        ILogger logger)
+        RazorSyntaxTree syntaxTree)
     {
         if (d.Range is null)
         {
             return false;
         }
 
-        var owner = syntaxTree.GetOwner(sourceText, d.Range.End, logger);
-        if (owner is null)
+        
+        var owner = syntaxTree.Root.FindNode(d.Range.AsRazorTextSpan(sourceText), getInnermostNodeForTie: true);
+        if (IsCsharpKind(owner))
         {
-            return false;
+            return true;
         }
 
-        var isCSharp = owner.Kind is SyntaxKind.CSharpExpressionLiteral or SyntaxKind.CSharpStatementLiteral or SyntaxKind.CSharpEphemeralTextLiteral;
+        if (owner is CSharpImplicitExpressionSyntax implicitExpressionSyntax &&
```

Comment:
I'm sure this could have been written with some fancy matching, but I couldn't figure it out for the 2 minutes I tried.

Reply:
I can't actually comment on the entire range, this replaces from here to `return false;`

```cs
return owner is CSharpImplicitExpressionSyntax { Body: CSharpImplicitExpressionBodySyntax { CSharpCode: CSharpCodeBlockSyntax { Children: [var child] } } }
       && IsCSharpKind(child);
```

Reply:
Oh, and I see that David had a separate comment with exactly that below.